File: windows/Services/SnippetHotKeyManager.cs
using WSNH.Models;

namespace WSNH.Services;

/// <summary>Registers/unregisters global hotkeys for each saved Snippet.</summary>
public sealed class SnippetHotKeyManager
{
    public static readonly SnippetHotKeyManager Shared = new();

    private readonly Dictionary<Guid, int> _registeredIds = new();
    public Action<Snippet>? OnTrigger { get; set; }

    private SnippetHotKeyManager() { }

    public void RegisterAll(IEnumerable<Snippet> snippets)
    {
        foreach (var id in _registeredIds.Values)
        {
            HotKeyRegistry.Shared.Unregister(id);
        }
        _registeredIds.Clear();

        foreach (var snippet in snippets)
        {
            Register(snippet);
        }
    }

    public void Register(Snippet snippet)
    {
        if (!snippet.HasHotkey) return;

        var registeredId = HotKeyRegistry.Shared.Register(snippet.KeyCode, snippet.ModifierFlags, () =>
        {
            OnTrigger?.Invoke(snippet);
        });

        if (registeredId is int id)
        {
            _registeredIds[snippet.Id] = id;
        }
    }

    public void Unregister(Guid snippetId)
    {
        if (_registeredIds.TryGetValue(snippetId, out var id))
        {
            HotKeyRegistry.Shared.Unregister(id);
            _registeredIds.Remove(snippetId);
        }
    }
}

File: mac/Sources/WSNH/Services/SnippetHotKeyManager.swift
import AppKit
import HotKey

/// Registers/unregisters global hotkeys for Snippets that have one set,
/// mirroring HotKeyManager but for direct-insert Snippets rather than
/// AI-rewrite PromptActions. Snippets without a hotkey (typed-shortcut-only)
/// are simply skipped here.
final class SnippetHotKeyManager {
    static let shared = SnippetHotKeyManager()

    private var hotKeys: [UUID: HotKey] = [:]
    var onTrigger: ((Snippet) -> Void)?

    private init() {}

    func registerAll(_ snippets: [Snippet]) {
        hotKeys.removeAll()
        for snippet in snippets where snippet.hasHotkey {
            register(snippet)
        }
    }

    func register(_ snippet: Snippet) {
        guard let key = Key(carbonKeyCode: snippet.keyCode) else { return }
        let modifiers = NSEvent.ModifierFlags(rawValue: snippet.modifierFlags)
        let hotKey = HotKey(key: key, modifiers: modifiers)
        hotKey.keyDownHandler = { [weak self] in
            self?.onTrigger?(snippet)
        }
        hotKeys[snippet.id] = hotKey
    }

    func unregister(_ snippetId: UUID) {
        hotKeys.removeValue(forKey: snippetId)
    }
}
